Add config_database.py

alltilla opened the pull request:
Usage:                                                                      
`./config_database.py <context> <driver>`

By default it caches the option database. Use the `-r` or `--rebuild` flag to force rebuild it.

For example:
```
$ ./config_database.py destination kafka-c
destination kafka-c(
  bootstrap-servers/kafka-bootstrap-servers(<string>)
  client-lib-dir(<string>)
  config/option()
  config/option(<string> <arrow> <string-or-number>)
  config/option(<string> <string-or-number>)
  flush-timeout-on-reload(<number>)
  flush-timeout-on-shutdown(<number>)
  frac-digits(<number>)
  key(<string>)
  local-time-zone/time-zone(<string>)
  log-fifo-size(<number>)
  message/template(<string>)
  on-error(<string>)
  persist-name(<string>)
  poll-timeout(<number>)
  properties-file(<path>)
  send-time-zone(<string>)
  sync-send(<yesno>)
  throttle(<number>)
  time-zone(<string>)
  topic(<string>)
  ts-format(<string>)
  workers(<number>)
  config/option(
    <string>(<string-or-number>)
  )
  key(
    <identifier>(<string>)
  )
  message/template(
    <identifier>(<string>)
  )
)
```

TODO:
- [x] Get types from grammar
- [x] ~Fix token resolve of same keyword in different drivers~ (Will be fixed in another PR)

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

furiel commented on config_option_database/config_helper.py:
```diff
@@ -0,0 +1,84 @@
+
+from pathlib import Path
+import json
+
```
config_helper.py sounds strange to me. config_database.py?

alltilla replied:
I am okay with renaming it! :)

alltilla commented on modules/afmongodb/afmongodb-grammar.ym:
```diff
@@ -88,11 +88,7 @@ afmongodb_option
 afmongodb_legacy_option
-    : KW_IFDEF
-        {
-#if SYSLOG_NG_ENABLE_LEGACY_MONGODB_OPTIONS
```
I removed this, because there is no way to access this rule, `KW_IFDEF` does not have a string representation in our `parser.c` files. It cannot be used currently, dead code. We can add a string representation in a `parser.c` file, if we intend to use it.

furiel replied:
I don't think the purpose of KW_IFDEF is to be reachable. It is just a dummy token that is intentionally cannot be reached from config. Its purpose is to wrap around the compiler switch that ends up in the generated c file. The compiler switch switches out a block of code, even if KW_IFDEF is not runtime-reachable. It is a compile time logic.

alltilla replied:
Wow, magic! :D I see now, thanks for the explanation. I will remove this change, and workaround this in another way.

alltilla replied:
I fixed it with `BisonGraph:get_parents(node)` and `ConfigOptions:_remove_ifdef(graph)`.

alltilla commented on config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,47 @@
+
+def _yacc2xml(yacc):
+    with _write_to_file(yacc) as file:
+        output = '/tmp/yacc2xml.xml'
```
`bison` takes a `.y` file (NamedTemporaryFile from `_write_to_file()`) and outputs a `.xml` file, which is our output.

furiel replied:
Oh ok I see.

Why do you use temporary file for yacc, but hardcoded file for the xml output?

alltilla replied:
I will change it to `output = NamedTemporaryFile().name`. It will close and delete the file, but generate a valid filename under `/tmp/`. After this, `bison` will generate the xml file at that location, and later we will delete that in `_yacc2rules()`.

alltilla replied:
Done!

alltilla commented on config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,47 @@
+        output = '/tmp/yacc2xml.xml'
+        try:
+            if not _run_in_shell(['bison', '--xml='+output, '--output=/dev/null', file.name]):
+                raise Exception('Failed to convert to xml:\n{}\n'.format(yacc))
```
It is not necessary, but we have a lot of grammar warnings, which I wanted to hide. I can show them. :)

furiel replied:
I understand your motivation now. If the failure cannot come from user error, which is apparently the case, I agree this can remain silence.

alltilla commented on config_option_database/utils/test/test_Yacc2Graph.py:
```diff
@@ -0,0 +1,62 @@
+
+def test_failed_yacc2xml():
+    with pytest.raises(Exception) as e:
+        _yacc2xml(test_string[2:])
```
We drop 2 characters to make an invalid yacc syntax, first is newline, second makes it invalid.

furiel replied:
Please `wrap test_string[2:]`.
For example: `generate_invalid_input(test_string)`.

Or even better, this would be enough?
```
 _yacc2xml("somethingthatisinvalid")
```
?

alltilla replied:
Good idea, I will use your second tip.

alltilla replied:
Done!

alltilla commented on tests/copyright/policy:
```diff
@@ -70,6 +71,7 @@ modules/http/(http|http-parser|http-plugin|)\.[ch]
 modules/http/http-grammar.ym
 modules/timestamp/rewrite-.*$
  GPLv2+_SSL
+config_option_database
```
I will move it :)

alltilla commented on config_option_database/utils/Yacc2Graph.py:
```diff
@@ -65,3 +66,9 @@ def _xml2rules(filename):
+
+def _yacc2rules(yacc):
+    xml = _yacc2xml(yacc)
+    rules = _xml2rules(xml)
```
Good idea, I will fix this, thanks!

alltilla replied:
Actually it is not hard to follow the cycle, the file (yacc) from `_write_to_file()` is only used in `_yacc2xml()` and it is removed there implicitly. The xml file is generated by bison and removed in `_yacc2rules()`. I think you thought, the 2 files are the same.

furiel replied:
Sorry for nitpicking, but then it would be preferable move `_write_to_file` definition into `_yacc2xml`. Their logic is heavily coupled.

alltilla replied:
I think it is cleaner this way.

alltilla commented on config_option_database/utils/Yacc2Graph.py:
```diff
@@ -45,3 +52,16 @@ def _yacc2xml(yacc):
+        number = int(rule.get('number'))
+        parent = rule.find('lhs').text
+        symbols = []
+        for symbol in rule.find('rhs'):
```
Thanks great idea! A small change is needed, though:
`symbols = [symbol.text for symbol in rule.find('rhs') if symbol.tag != 'empty']`

alltilla commented on config_option_database/utils/Yacc2Graph.py:
```diff
@@ -72,3 +74,15 @@ def _yacc2rules(yacc):
+    graph = networkx.DiGraph()
+    for rule in rules:
+        rule_node = str(rule.number)
+        graph.add_edge(rule.parent, rule_node)
```
If this is just a rename, I can rename it :)

alltilla replied:
Oh wait, it is a `networkx` function. Isn't arc the same as edge? I did not find arc in `networkx`.

furiel replied:
I did not find that either. Though we are building an api around networkx, and we can choose our own terminology. Strictly speaking, there is no edge in a directed graph. Well there is, but it is defined as two arcs with opposing direction.

Again this is an optional comment. :)

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -0,0 +1,42 @@
+        return list(self.graph.nodes)
+
+    def is_junction(self, node):
+        if node not in self.get_nodes():
```
I think the naming is inconvenient here. This could be `is_rule()`, as in the commit message of 5ac8d22f1168fba1cf03fde5577f4073ad570062 .

furiel replied:
Do you plan to rename?

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -32,6 +32,15 @@ def __init__(self, yaccfile):
 
+    def get_children(self, node):
+        children_as_list = self.graph.successors(node)
+        if children_as_list and self.is_junction(node):
```
Same here, `is_junction()` could be `is_rule()`. Rules' children must be sorted by `index`. `successors()` does not return the `index` of the edge/arc, but `self.graph[node].items()` does.
I will reorganize this function for sure, as it is not really intuitive right now :)

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -32,6 +32,15 @@ def __init__(self, yaccfile):
+    def get_children(self, node):
+        children_as_list = self.graph.successors(node)
+        if children_as_list and self.is_junction(node):
+            children = [(k, v) for k, v in self.graph[node].items()]
```
Sure! :)

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -60,3 +60,33 @@ def make_terminal(self, node):
+                if self.is_terminal(child):
+                    if child == '$end':
+                        break
+                    paths = paths[:]
```
The function gets `paths` as an argument, and returns `paths`. Although the naming is unfortunate, it is intentional, to not change the `paths` in the argument of the function, because it would have an effect on other rules. Easiest is to comment it out, and run the `pytest`. But the copy should not be done for every child, so it is better to move it to the beginning of the function, so it only runs once.

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -60,3 +60,33 @@ def make_terminal(self, node):
     def remove(self, node):
         self.graph.remove_node(node)
+
+    def get_paths(self, node='$accept', paths=None, stack=None):
```
Yeah, you are right, it is `$accept`

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -60,3 +60,33 @@ def make_terminal(self, node):
+                    for i in range(len(paths)):
+                        paths[i] += (child,)
+                else:
+                    paths = self.get_paths(child, paths, stack)
```
The joining is done in line 88, deeper in the recursion.

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -60,3 +60,33 @@ def make_terminal(self, node):
         self.graph.remove_node(node)
+
+    def get_paths(self, node='$accept', paths=None, stack=None):
+        if stack is None:
```
Yes, it is, and it can be a `set()`.

alltilla commented on config_option_database/utils/BisonGraph.py:
```diff
@@ -60,3 +60,33 @@ def make_terminal(self, node):
+                    if child == '$end':
+                        break
+                    paths = paths[:]
+                    for i in range(len(paths)):
```
We only have paths here, which contain the current node.

furiel replied:
Oh you are right. Overlook from my side..

alltilla commented:
@furiel Thanks for the review, I answered/fixed everything you mentioned! :)

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build SUCCESS

alltilla commented:
Added options from `.scl` files.

kira-syslogng commented:
Build SUCCESS

alltilla commented:
@furiel: Sorry, I pushed some code again :( Let's discuss the review/PR update process for this PR on Monday, so I won't make you this big of a trouble again.

kira-syslogng commented:
Build SUCCESS

alltilla commented on tests/copyright/policy:
```diff
@@ -70,6 +71,7 @@ modules/http/(http|http-parser|http-plugin|)\.[ch]
 modules/http/http-grammar.ym
 modules/timestamp/rewrite-.*$
  GPLv2+_SSL
+contib/config_option_database
```
Yes, you are absolutely right.

alltilla replied:
Fixed!

alltilla commented on contrib/config_option_database/utils/ConfigOptions.py:
```diff
@@ -0,0 +1,123 @@
+        graph.remove(to_node)
+
+def _remove_code_blocks(graph):
+    for node in filter(lambda x: x.startswith('$@'), graph.get_nodes()):
```
Sure!
Take this grammar for example:
```
http_destination
    : KW_HTTP
      {
        last_driver = http_dd_new(configuration);
      }
      '(' http_options ')'  { $$ = last_driver; }
    ;
```

`bison` converts this to:
```
<rule number="128" usefulness="useful">
  <lhs>http_destination</lhs>
  <rhs>
    <symbol>KW_HTTP</symbol>
    <symbol>$@16</symbol>
    <symbol>'('</symbol>
    <symbol>http_options</symbol>
    <symbol>')'</symbol>
  </rhs>
</rule>
```
The `last_driver = http_dd_new(configuration);` code is converted to `$@16`, which is unnecessary for us. These are what I remove with this function.

alltilla commented on contrib/config_option_database/utils/ConfigOptions.py:
```diff
@@ -0,0 +1,123 @@
+def get_driver_options():
+    return _get_driver_options_from_grammar() | _get_driver_options_from_scl()
+
+def _merge_option_property_style(add_to, option):
```
I don't see this code to be motivated by Light. I would like to make more use cases for it. `config_database.py` itself is another use case, and hopefully more will come.

alltilla replied:
I misunderstood you. Dropped the patch :)

alltilla commented on contrib/config_option_database/utils/ConfigOptions.py:
```diff
@@ -0,0 +1,123 @@
+    return _get_driver_options_from_grammar() | _get_driver_options_from_scl()
+
+def _merge_option_property_style(add_to, option):
+    for other in add_to:
```
Thanks, fixed!

alltilla commented on contrib/config_option_database/utils/BisonGraph.py:
```diff
@@ -0,0 +1,104 @@
+            stack = set()
+        if paths is None:
+            paths = [()]
+        paths = paths[:]
```
Yes, sure, done.

alltilla commented on contrib/config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,85 @@
+    remove(xml)
+    return rules
+
+def _rules2graph(rules):
```
Sure, added! :)

kira-syslogng commented:
Build SUCCESS

alltilla commented:
#making_travis_commits_check_happy

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE

alltilla commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

gaborznagy commented:
I've quickly checked the patches, I made some notes on which I have to go through.

I have 2 remarks now, I'm interested in your opinion:
- a feature request, strictly for a next PR:  list drivers
  Either by giving only the context as paremeter `config_database.py destination | grep elastic` and "driver" is optional,  or with an explicit option --list-drivers.
- including the unit tests in Travis CI
I don't know which rule should we use, or a new one, but it would be useful.

alltilla commented:
@furiel Squashed the FIXUP commit, and added some whitespace changes based on `flake8`.

alltilla commented:
@gaborznagy 
> I've quickly checked the patches, I made some notes on which I have to go through.
> 
> I have 2 remarks now, I'm interested in your opinion:
> 
> * a feature request, strictly for a next PR:  list drivers
>   Either by giving only the context as paremeter `config_database.py destination | grep elastic` and "driver" is optional,  or with an explicit option --list-drivers.
> * including the unit tests in Travis CI
>   I don't know which rule should we use, or a new one, but it would be useful.

Yes, more options of `config_database.py` will come in following patches, thanks for the tip!
Regarding the travis, I would like to create a `make` rule for all python unit tests, including `Light` and this, too. That rule could be ran in a separate travis task, I think.

kira-syslogng commented:
Build SUCCESS

mitzkia commented on contrib/config_option_database/utils/ConfigOptions.py:
```diff
@@ -0,0 +1,72 @@
+            graph.remove(parent)
+
+
+def get_driver_options():
```
I think the scope of this function is more than only to getting "driver" "options".
This function could explore not just: drivers, but all grammar elements: junctions, channel, parsers, rewrite rules, so every statement which connects to grammar.
And explores not just options, but also blocks (I mean the whole grammar hierarchy).
Please consider to rename this function.

alltilla replied:
I don't plan on handling the junctions or channels here. Parsers, rewrites will come here eventually, so we can change the "driver" in the function name. Any suggestions?

mitzkia replied:
What about `get_options_for_statements` or `get_options_by_statements` ?

mitzkia replied:
What about `get_options_by_statements_and_blocks` ?

mitzkia commented on contrib/config_option_database/utils/MergeYm.py:
```diff
@@ -0,0 +1,70 @@
+
+
+def skip_filepath(filepath):
+    return 'plugin_template_grammar.ym' == filepath.name
```
`return filepath.name in ['plugin_template_grammar.ym']` could be more future proof form

alltilla replied:
Removed it altogether with the `modules` dir patch. Thanks :)

gaborznagy commented on contrib/config_option_database/utils/MergeYm.py:
```diff
@@ -0,0 +1,70 @@
+
+def get_grammar_files():
+    root_dir = Path(__file__).resolve().parents[3]
+    files = list(root_dir.rglob('*.ym'))
```
I'd answer it, since I just left here a comment: 
good question, but the generated grammar files have '.y' extension, e.g. `build/modules/afsocket/afsocket-grammar.y`.

However, it would be good to include only `modules` dir, or exclude the file `./dev-utils/plugin_skeleton_creator/plugin_template_grammar.ym`.

alltilla replied:
Done, good idea, thanks! :)

gaborznagy commented on contrib/config_option_database/utils/MergeYm.py:
```diff
@@ -0,0 +1,70 @@
+
+
+def get_grammar_files():
+    root_dir = Path(__file__).resolve().parents[3]
```
This is very sensitive to the module's file path. If it's relocated it won't work.

I see that it's hard to inject a root path, I've looked around in our repository and found that some python functional tests are "cheating": they rely on the `top_builddir` env variable:
`tests/functional/globals.py:    for (root, dirs, files) in os.walk(os.path.abspath(os.path.join(os.environ['top_builddir'], 'modules'))):`

Since we don't install the grammar files, this script has to be used in the source repository.
Just after a quick search, I've found that you can ask the git repository work tree's path with `git rev-parse --show-toplevel`. Maybe this can be injected from outside.

But this can be changed in a subsequent PR.

alltilla replied:
The tarball, we provide as release package is not packed with git folder, so `git rev-parse` might not be the best solution.

furiel replied:
Emacs has a `locate-dominating-file` function which looks for a specific file name recursively upwards the directory hierarcy, either until the file is found or / directory is reached. Returns with the directory that contains the file name. Similar probing could be implemented, for example you can use `VERSION` or any other file in the project-root directory.

alltilla replied:
Thanks, I will try that!

gaborznagy replied:
> The tarball, we provide as release package is not packed with git folder, so git rev-parse might not be the best solution.

Since we don't provide the grammar files in the release tarball, you cannot use the script at all in a release tarball, so that does not exclude git rev-parse.

@furiel's idea can work, but it has the possibility that the script may return with the wrong path if it does not found the marker-file (`VERSION` was mentioned).
With a good error-handling (root dir is not allowed), this can be robust.
Maybe the file `syslog-ng-config.h.in` would be a more project-specific file, but there can be additional markers as well: in case the first one (VERSION) is found, it is expected to that the second marker (syslog-ng-config.h.in) is next to it.

alltilla commented on contrib/config_option_database/config_database.py:
```diff
@@ -0,0 +1,91 @@
+    db = {'source': {}, 'destination': {}}
+    for context, driver, keyword, arguments, parents in get_driver_options():
+        db[context].setdefault(driver, {'options': [], 'blocks': {}})
+        add_to = db[context][driver]
```
It is purposely general. It changes with the parents of the option, so it does not stay a `driver_node` if the option has parents.

alltilla replied:
With this context, could you please suggest another name? I couldn't think of one :)

mitzkia replied:
What about `database_options_block` or `database_options_node` or `database_options_level`? 
It is not easy, because we are jumping between options normal level and options block levels.

alltilla commented:
@gaborznagy 
> I think config_option_database files should be added in the dist tarball (`EXTRA_DIST`).
> 
> I've just realized that we don't install the grammar files, which means this cannot be used without the source repository.
> Just a hint for future: we can package the output database with the dist tarball.

We don't install them, but we put them in the release package, so `EXTRA_DIST` is okay for now.
Yes the ideal solution will be to install the database and the python script, too.

alltilla commented on contrib/config_option_database/utils/test/test_OptionParser.py:
```diff
@@ -0,0 +1,247 @@
+
+
+def test_find_options_with_keyword():
+    tests = [
```
Done!

alltilla commented on contrib/config_option_database/utils/OptionParser.py:
```diff
@@ -0,0 +1,153 @@
+        resolve_db = {}
+        struct_regex = re.compile(r'CfgLexerKeyword[^;]*')
+        entry_regex = re.compile(r'{[^{}]+,[^{}]+}')
+        for f in Path(__file__).parents[3].rglob('*-parser.c'):
```
I moved it to a variable for now. We can move it to a function, when we came up with an idea to generalize it :)

alltilla commented on contrib/config_option_database/utils/BisonGraph.py:
```diff
@@ -0,0 +1,99 @@
+        stack.add(node)
+
+        if self.is_rule(node):
+            for child in self.get_children(node):
```
Done, thanks! :)

alltilla commented on contrib/config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,131 @@
+    return file
+
+
+def _yacc2xml(yacc):
```
Done, thanks! :)

alltilla commented on contrib/config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,131 @@
+
+def _yacc2xml(yacc):
+    with _write_to_file(yacc) as file:
+        output = NamedTemporaryFile().name
```
Renamed it to `xml_filepath`, because it suits it better, I believe. :)
Here we generate a temp file name, and ask bison to generate the xml-file there, so it is not temporary. It will be removed later explicitly.

alltilla commented:
@furiel @mitzkia @gaborznagy :
Thank you for the good comments, I updated the branch with FIXUP, REBASED, NEW commits, these are all you have to check now. :)

alltilla commented on contrib/config_option_database/utils/Yacc2Graph.py:
```diff
@@ -0,0 +1,131 @@
+    return proc.returncode == 0
+
+
+def _write_to_file(string):
```
Done, thanks! :)

kira-syslogng commented:
Build FAILURE

kira-syslogng commented:
Build FAILURE

alltilla commented:
Rebased to latest-greatest master, and squashed the fixup commits :)

kira-syslogng commented:
Build SUCCESS

mitzkia commented on contrib/config_option_database/utils/test/test_OptionParser.py:
```diff
@@ -0,0 +1,238 @@
+    ]
+)
+def test_path_to_options(path, options):
+    assert path_to_options(path.split()) == options
```
Turned out that this test is flaky.

```python
    def test_path_to_options(path, options):

>       assert path_to_options(path.split()) == options

E       AssertionError: assert [('destinatio... 'failback'))] == [('destinatio...ring>',), ())]

E         At index 0 diff: ('destination', 'udp', '', ('<string>',), ()) != ('destination', 'udp', 'servers', ('<string-list>',), ('failover',))

E         Use -v to get the full diff

utils/test/test_OptionParser.py:238: AssertionError

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB post_mortem (IO-capturing turned off) >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

> /home/micek/git_projects/github_mitzkia/syslog-ng/contrib/config_option_database/utils/test/test_OptionParser.py(238)test_path_to_options()

-> assert path_to_options(path.split()) == options

(Pdb) p options

[('destination', 'udp', 'servers', ('<string-list>',), ('failover',)), ('destination', 'udp', 'tcp-probe-interval', ('<positive-integer>',), ('failover', 'failback')), ('destination', 'udp', '', ('<string>',), ())]

(Pdb) p path_to_options(path.split())

[('destination', 'udp', '', ('<string>',), ()), ('destination', 'udp', 'servers', ('<string-list>',), ('failover',)), ('destination', 'udp', 'tcp-probe-interval', ('<positive-integer>',), ('failover', 'failback'))]
```

alltilla replied:
Good catch, thanks! Fixed :)

mitzkia commented on contrib/config_option_database/utils/OptionParser.py:
```diff
@@ -0,0 +1,154 @@
+    resolved_tokens = []
+    for token in tokens:
+        resolved_tokens.append(_resolve_token(token))
+    return tuple(resolved_tokens)
```
Is converting to `tuple` is necessary here? If it so than, why not initialized original resolved_tokens as `tuple`?

alltilla replied:
You are right, fixed! :)

kira-syslogng commented:
Build FAILURE

alltilla commented:
@kira-syslogng retest this please

kira-syslogng commented:
Build SUCCESS

alltilla commented:
rebased to master :)

kira-syslogng commented:
Build SUCCESS

kira-syslogng commented:
Build FAILURE
